### PR TITLE
Wada basins detection with the merging algorithm

### PR DESCRIPTION
**Pull request by awage**
This is an implementation of the merging algorithm to detect Wada basins. EDIT: only possible in 2D

It is fast and simple to understand. The distance between boundaries is computed using the distance transform of a binary matrix. See this thread: 

https://cs.stackexchange.com/questions/117989/hausdorff-distance-between-two-binary-images-according-to-distance-maps

The boundaries can also be computed using this distance matrix so the code is minimal and quite fast (a few ms for 1000x1000 basins matrix) 

A. Daza, Alexandre Wagemakers and Miguel A.F. Sanjuán. Ascertaining when a basin is Wada: The merging method. Scientific Reports 8, 9954, 2018.

The basic idea of the algorithm is to compute all boundaries between a basin and the other merged basins. If all boundaries are within some distance r in pixels then we judge that the basins are Wada: there is a unique boundary. 

The Doc still needs improvement and the `haussdorff_distance`  is optional. Tests are based on an artificial basins matrix.

### CODE REVIEW

File: src/basins/wada_test.jl [hunk 1]
```diff
@@ -0,0 +1,161 @@
+export test_wada_merge, haussdorff_distance
+
+# Generate all pairs of the number in ids without 
+# repetition
+function generate_pairs(ids) 
+ p = Vector{Tuple{Int}}()
```
Comment (Datseris) [inline src/basins/wada_test.jl:6]: `Tuple{Int}` is not the elements of the container. `Tuple{Int, Int}` is.

File: src/basins/wada_test.jl [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    c1 = count(bd1.*bnd2 .> r)
+    c2 = count(bd2.*bnd1 .> r)
+    # @show c1, c2
+    return max(c1,c2)./length(bnd1)
+end
+
+"""
+   merged_boundary(basins, id) -> basins
+
+Returns a matrix with two basins: the first is the basins
```
Comment (Datseris) [inline src/basins/wada_test.jl:105]: Return instead of Returns

File: src/basins/wada_test.jl [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        mrg_basins[I] .= 1
+    end
+    return mrg_basins
+end
+
+"""
+    test_wada_merge(basins, r) -> p
+
+This function test if the 2D array `basins` has the Wada 
+property using the merging technique [Daza2018](@cite).
```
Comment (Datseris) [inline src/basins/wada_test.jl:125]:
```suggestion
Test if the 2D array `basins` has the Wada 
property using the merging technique of [Daza2018](@cite). 
```

File: src/basins/wada_test.jl [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+these two objects. The algorithm then computes the distance between 
+each boundaries of these basins pairwise. If all the boundaries are 
+within some distance `r`, there is a unique boundary separating 
+the basins and we have the wada property. 
+The algorithm returns the maximum proportion of pixels of a boundary 
+with distance strictly greater than `r` from another boundary. 
+
+If `p == 0`,  we have the Wada property for this value of `r`. 
+If `p > 0`, the criteria to decide if
+the basins are Wada is left to the researcher. Numerical inaccuracies
```
Comment (Datseris) [inline src/basins/wada_test.jl:141]:
```suggestion
the basins are Wada is left to the user. Numerical inaccuracies 
```

File: src/basins/wada_test.jl (hunk 3 above)
Comment (Datseris) [reply src/basins/wada_test.jl:125]: Addition, I perhpas it would be useful to cite a paper or Wiki entry that defines the "Wada property".

**Review by Datseris — COMMENTED** [review]
I have some comments for re-organizing the source code. Furthermore, tests fail. Is the test failure related with this PR? please check. and if not i'll try to fix the tests in another pr.

File: src/basins/wada_test.jl
[the remark this answers is not in the document]
Comment (awage) [reply src/basins/wada_test.jl:1]: Got it!

File: src/basins/wada_test.jl [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+end
+
+
+# Function for L1 metric 
+w(M) = min(M[1,2]+1, M[2,1]+1, M[2,2]) 
+w2(M) = min(M[1,2]+1, M[2,1]+1, M[1,1]) 
+
+# Function for L∞ metric (Chebyshev)
+# w(M) = min(M[1,2]+1, M[2,1]+1, M[2,2], M[1,1] +1) 
+# w2(M) = min(M[1,2]+1, M[2,1]+1, M[1,1], M[2,2] +1)
```
[the remark this answers is not in the document]
Comment (awage) [reply src/basins/wada_test.jl:36]: This is a possible extension of the functionalities. The L_inf metric can also be used but it is not critical. I will remove it.

**Comment by Datseris** [comment]
@awage have you checked where the test failure is coming from?

**Comment by awage** [comment]
I think the problem comes from the docs. All tests are passing locally but there is an error when I try to build the docs. 

```
 Error: no docs found for 'reinit!(::DynamicalSystem, args...; kwargs...)' in `@docs` block in src/dynsysref.md:32-36
│ ```@docs
│ step!(::DynamicalSystem, args...; kwargs...)
│ reinit!(::DynamicalSystem, args...; kwargs...)
│ set_parameter!
│ ```

```

**Comment by awage** [comment]
And also: 

```julia
ERROR: LoadError: `makedocs` encountered an error [:docs_block] -- terminating build before rendering.
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
  [2] runner(::Type{Documenter.Builder.RenderDocument}, doc::Documenter.Document)
    @ Documenter ~/.julia/packages/Documenter/CJeWX/src/builder_pipeline.jl:253
  [3] dispatch(::Type{Documenter.Builder.DocumentPipeline}, x::Documenter.Document)
    @ Documenter.Selectors ~/.julia/packages/Documenter/CJeWX/src/utilities/Selectors.jl:170
  [4] #86
    @ ~/.julia/packages/Documenter/CJeWX/src/makedocs.jl:248 [inlined]
  [5] withenv(::Documenter.var"#86#88"{Documenter.Document}, ::Pair{String, Nothing}, ::Vararg{Pair{String, Nothing}})
    @ Base ./env.jl:257
  [6] #85
    @ ~/.julia/packages/Documenter/CJeWX/src/makedocs.jl:247 [inlined]
  [7] cd(f::Documenter.var"#85#87"{Documenter.Document}, dir::String)
    @ Base.Filesystem ./file.jl:112
  [8] makedocs(; debug::Bool, format::Documenter.HTMLWriter.HTML, kwargs::@Kwargs{…})
    @ Documenter ~/.julia/packages/Documenter/CJeWX/src/makedocs.jl:247
  [9] makedocs
    @ ~/.julia/packages/Documenter/CJeWX/src/makedocs.jl:241 [inlined]
 [10] build_docs_with_style(::Vector{…}, ::Module, ::Vararg{…}; bib::CitationBibliography, authors::String, draft::Bool, htmlkw::@NamedTuple{}, kwargs::@Kwargs{…})
    @ Main ~/urjc/investigacion/proyecto_basins_computing/Attractors.jl/docs/build_docs_with_style.jl:78
 [11] top-level scope
    @ ~/urjc/investigacion/proyecto_basins_computing/Attractors.jl/docs/make.jl:31
 [12] include(fname::String)
    @ Base.MainInclude ./client.jl:489
 [13] top-level scope
    @ REPL[2]:1

```

**Review by Datseris — APPROVED** [review]
Okay, please add a changelog entry and merge. I am taking care of the docs in my pr #125

**Pull request merged**